### PR TITLE
Add Microsoft.VisualBasic source

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
cc @joshfree, @Priya91, @mellinoe, @stephentoub

File: dir.props [hunk 1]
```diff
@@ -66,13 +66,13 @@
     <DebugSymbols Condition="'$(DebugSymbols)' == ''">true</DebugSymbols>
     <Optimize Condition="'$(Optimize)' == ''">false</Optimize>
     <DebugType Condition="'$(DebugType)' == ''">full</DebugType>
-    <DefineConstants>$(DefineConstants);DEBUG;TRACE</DefineConstants>
+    <DefineConstants>$(DefineConstants),DEBUG,TRACE</DefineConstants>
```
Comment (stephentoub) [inline dir.props]: Why the change to commas?

File: src/Microsoft.VisualBasic/src/Microsoft.VisualBasic.vbproj [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    <Compile Include="Microsoft\VisualBasic\CompilerServices\Utils.LateBinder.vb" />
+    <Compile Include="Microsoft\VisualBasic\CompilerServices\Utils.vb" />
+    <Compile Include="Microsoft\VisualBasic\Constants.vb" />
+    <Compile Include="Microsoft\VisualBasic\ControlChars.vb" />
+    <Compile Include="Microsoft\VisualBasic\HideModuleNameAttribute.vb" />
+    <Compile Include="Microsoft\VisualBasic\Interaction.vb" />
+    <Compile Include="Microsoft\VisualBasic\Strings.vb" />
+  </ItemGroup>
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.targets))\dir.targets" />
+</Project>
```
Comment (stephentoub) [inline src/Microsoft.VisualBasic/src/Microsoft.VisualBasic.vbproj]: Nit:some of the files need newlines at the end.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/ControlChars.vb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+' Copyright (c) Microsoft. All rights reserved.
+' Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+Namespace Microsoft.VisualBasic
+    ' Contants for the Control Characters
+    Public NotInheritable Class ControlChars
+
+        Public Const CrLf As String = ChrW(13) & ChrW(10)
+        Public Const NewLine As String = ChrW(13) & ChrW(10)
```
Comment (stephentoub) [inline src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/ControlChars.vb]: Should NewLine be defined in terms of Environment.NewLine?  The latter will be "\r\n" on Windows but "\n" on Unix, and I assume we'd want the same behavior for ControlChars.NewLine?

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Constants.vb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+﻿' Copyright (c) Microsoft. All rights reserved.
+' Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+Namespace Global.Microsoft.VisualBasic
+    <Global.System.Diagnostics.DebuggerNonUserCode()>
+    Public Module Constants
+        Public Const vbCrLf As String = ChrW(13) & ChrW(10)
+        Public Const vbNewLine As String = ChrW(13) & ChrW(10)
```
Comment (stephentoub) [inline src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Constants.vb]: Same comment as for ControlChars.NewLine.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/CompilerServices/NewLateBinding.vb [hunk 5]
```diff
@@ -0,0 +1,1503 @@
...
+                        Dim errorMessage As String = ""
+                        Dim errors As New List(Of String)
+
+                        Dim result As Boolean =
+                            CanMatchArguments(resolutionResult, arguments, argumentNames, typeArguments, False, errors)
+
+                        Debug.Assert(result = False AndAlso errors.Count > 0, "expected this candidate to fail")
+
+                        For Each errorString As String In errors
+                            errorMessage &= vbCrLf & "    " & errorString
```
Comment (stephentoub) [inline src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/CompilerServices/NewLateBinding.vb]: There are a handful of places in the source where vbCrLf is being used to introduce newlines... should vbNewLine be used instead, especially if it's changed to be Environment.NewLine and have the right newline characters cross-plat?

File: src/Microsoft.VisualBasic/src/packages.config [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  <package id="System.Reflection.Primitives" version="4.0.0-beta-22703" targetFramework="portable-net45+win" />
+  <package id="System.Reflection.TypeExtensions" version="4.0.0-beta-22703" targetFramework="portable-net45+win" />
+  <package id="System.Resources.ResourceManager" version="4.0.0-beta-22703" targetFramework="portable-net45+win" />
+  <package id="System.Runtime" version="4.0.20-beta-22703" targetFramework="portable-net45+win" />
+  <package id="System.Runtime.Extensions" version="4.0.10-beta-22703" targetFramework="portable-net45+win" />
+  <package id="System.Runtime.Handles" version="4.0.0-beta-22703" targetFramework="portable-net45+win" />
+  <package id="System.Runtime.InteropServices" version="4.0.20-beta-22703" targetFramework="portable-net45+win" />
+  <package id="System.Text.Encoding" version="4.0.10-beta-22703" targetFramework="portable-net45+win" />
+  <package id="System.Threading" version="4.0.10-beta-22703" targetFramework="portable-net45+win" />
+  <package id="System.Threading.Tasks" version="4.0.10-beta-22405" targetFramework="portable-net45+win" />
```
Comment (stephentoub) [inline src/Microsoft.VisualBasic/src/packages.config]: Any reason 22703 isn't used for this version as with the others?

**Comment by stephentoub** [comment]
A few minor nits, but otherwise I skimmed through it and it looks fine to me.

File: dir.props [hunk 7]
```diff
@@ -2,7 +2,7 @@
   <Import Condition="Exists('..\dir.props')" Project="..\dir.props" />
   <!-- Build Tools Version -->
   <PropertyGroup>
-    <BuildToolsVersion>1.0.25-prerelease-00023</BuildToolsVersion>
+    <BuildToolsVersion>1.0.25-prerelease-00025</BuildToolsVersion>
```
Comment (Priya91) [inline dir.props]: the buildtools in the repo changed to 00026 (the latest), please undo this change while merging.

**Comment by bbowyersmyth** [comment]
What is the coding standard rule on the ByVal being explicitly declared in VB?

It seems like visual noise now and since VS2010 SP1 it is no longer automatically added for new code.

File: dir.props (hunk 1 above)
Comment (AlekseyTs) [inline dir.props]: VB compiler doesn't support ';' as a separator. C# compiler supports both.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/ControlChars.vb (hunk 3 above)
Comment (AlekseyTs) [inline src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/ControlChars.vb]: I do not know the answer at the moment. In any case this is out of scope for this commit.

File: src/Microsoft.VisualBasic/src/packages.config (hunk 6 above)
Comment (AlekseyTs) [inline src/Microsoft.VisualBasic/src/packages.config]: For some reason package manager in VS picked this version. I'll change this manually.

File: dir.props (hunk 7 above)
Comment (AlekseyTs) [inline dir.props]: Will do. Thanks.

**Comment by AlekseyTs** [comment]
@bbowyersmyth I do not believe ByVal is prohibited.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/ControlChars.vb (hunk 3 above)
Comment (rolfbjarne) [inline src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/ControlChars.vb]: @stephentoub: NewLine is a constant, and is embedded inside the assembly by the VB compiler. This means that if the value is platform-dependent ("\r\n" on Windows, but "\n" on Unix) you'll get different behavior depending on where you compiled your app, not depending on where you run it, which is not what most people (everybody?) expect.

Comment (stephentoub) [inline src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/ControlChars.vb]: @rolfbjarne, I agree, hence my question.  I've opened #1817.
